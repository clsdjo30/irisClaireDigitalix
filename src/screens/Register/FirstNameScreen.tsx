import React, { useState } from 'react';
import { StyleSheet, Text, View, Dimensions } from 'react-native';
import { Input, Icon } from '@rneui/base';
import { StackScreenProps } from '@react-navigation/stack';
import { useUserStore } from '../../hooks/useUserStore';
import { colors } from '../../theme';
import NavigationButton from '../../components/NavigationButton';

import { SCREEN_WIDTH } from "../../utils/constants";

const FirstNameScreen: React.FC<StackScreenProps<any>> = ({ navigation }) => {
  const [user, setUser] = useUserStore();
  const [error, setError] = useState('');

//interdire les caracteres speciaux
  const isValidName = (text: string) => {
    const regex = /^[A-Za-z]+$/;
    return regex.test(text);
  };

  // Controller la saisi du prenom
  const handleInputChange = (text: string) => {
    setError('');

    // Si le texte est vide, permettez la mise à jour
    if (text === '') {
      setUser({ ...user, firstname: text });
      return;
    }

    if (!isValidName(text)) {
      setError('Veuillez entrer uniquement des lettres.');
      return;
    }

    setUser({ ...user, firstname: text });
  };

  const goToGenreScreen = () => {
    if (user.firstname.length === 0) {
      setError('Le prénom ne peut pas être vide');
      return;
    }

    if (user.firstname.length < 3) {
      setError('Le prénom doit contenir au moins 3 caractères');
      return;
    }

    navigation.navigate('Genre');
  };

  return (
    <View testID='first-name-screen' style={styles.container}>
      {error && <View style={styles.errorContainer}><Text style={styles.errorText}>{error}</Text></View>}
      <View style={styles.controls}>
        <View style={styles.genderTitle}>
          <Text style={styles.contentTitle}>Quelle est votre prénom ?</Text>
        </View>
        <View style={styles.inputView}>
          <Input
            value={user.firstname}
            placeholder='Saisissez votre prénom'
            placeholderTextColor={colors.palette.purple200}
            style={styles.input}
            onChangeText={handleInputChange}
            leftIcon={<Icon
              name="mail-outline"
              type="ionicon"
              color={colors.palette.white}
              size={20}
            />}
          />
        </View>
        <NavigationButton
          color={colors.palette.violetBg}
          backgroundColor={colors.palette.orange}
          width={SCREEN_WIDTH * 0.85}
          title="Suivant"
          onPress={goToGenreScreen}
        />
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: colors.palette.stepViolet
  },
  controls: {
    flexDirection: 'column',
    alignItems: 'center',
    marginTop: 40,
  },
  inputView: {
    width: SCREEN_WIDTH * 0.9,
    marginBottom: 20
  },
  input: {
    width: SCREEN_WIDTH / 0.3,
    height: 48,
    backgroundColor: colors.palette.stepViolet,
    borderRadius: 16,
    paddingLeft: 20,
    fontFamily: "mulishExtraLight",
    fontSize: 18,
    color: colors.palette.white,
  },
  inputStyle: {
    fontSize: 16,
    marginLeft: 10,
    fontFamily: "mulishRegular",
    color: colors.palette.violet
  },
  icon: {
    marginLeft: 10,
    color: colors.palette.violet,
  },
  genderTitle: {
    width: 300,
    flexDirection: 'row',
  },
  contentTitle: {
    fontFamily: "mulishSemiBold",
    fontSize: 20,
    color: colors.palette.violetClair,
    marginBottom: 20
  },
  //ERROR
  errorContainer: {
    position: 'absolute',
    top: 0,
    width: '80%',
    backgroundColor: colors.palette.orange,
    marginTop: 40,
    borderRadius: 16,
  },
  errorText: {
    textAlign: "center",
    fontFamily: "mulishBold",
    fontSize: 12,
    color: colors.palette.violetClair,
    paddingVertical: 6,
  },

});

export default FirstNameScreen;