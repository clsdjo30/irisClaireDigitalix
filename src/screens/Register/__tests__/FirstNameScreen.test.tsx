import { render, fireEvent } from '@testing-library/react-native';
import FirstNameScreen from '../FirstNameScreen';

// Mock the navigation object with required methods
const mockNavigation: any = {
    navigate: jest.fn(),
};
const mockRoute: any = {
    key: 'MockKey',
    name: 'MockName',
};

// Mock the Icon component from react-native-vector-icons/FontAwesome
jest.mock('react-native-vector-icons/FontAwesome', () => 'Icon');

describe('FirstNameScreen', () => {
    it('renders correctly and matches snapshot', () => {
        const tree = render(<FirstNameScreen navigation={mockNavigation} route={mockRoute} />);
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly', () => {
        const { getByTestId } = render(<FirstNameScreen navigation={mockNavigation} route={mockRoute} />);
        expect(getByTestId('first-name-screen')).toBeTruthy();
    });

    it('contains the key elements', () => {
        const { getByText, getByPlaceholderText } = render(<FirstNameScreen navigation={mockNavigation} route={mockRoute} />);
        expect(getByText('Quelle est votre prénom ?')).toBeTruthy();
        expect(getByPlaceholderText('Saisissez votre prénom')).toBeTruthy();
    });

    it('updates the user store when the input changes', () => {
        const { getByPlaceholderText } = render(<FirstNameScreen navigation={mockNavigation} route={mockRoute} />);
        const input = getByPlaceholderText('Saisissez votre prénom');
        fireEvent.changeText(input, 'John');
        // Add a check to verify that user.firstname in the store is now 'John'
    });

    it('navigates to the next screen when the Next button is pressed', () => {
        const { getByText, getByPlaceholderText } = render(<FirstNameScreen navigation={mockNavigation} route={mockRoute} />);

        // Simulate a valid input
        const input = getByPlaceholderText('Saisissez votre prénom');
        fireEvent.changeText(input, 'John');

        const nextButton = getByText('Suivant');
        fireEvent.press(nextButton);

        expect(mockNavigation.navigate).toHaveBeenCalledWith('Genre');
    }); 

    it('updates the user state when the input changes', () => {
        const { getByPlaceholderText } = render(<FirstNameScreen navigation={mockNavigation} route={mockRoute} />);
        const input = getByPlaceholderText('Saisissez votre prénom');
        fireEvent.changeText(input, 'John');
    });

    it('renders key elements correctly', () => {
        const { getByText, getByPlaceholderText } = render(<FirstNameScreen navigation={mockNavigation} route={mockRoute} />);
        expect(getByText('Quelle est votre prénom ?')).toBeTruthy();
        expect(getByPlaceholderText('Saisissez votre prénom')).toBeTruthy();
        expect(getByText('Suivant')).toBeTruthy();
    });


});
