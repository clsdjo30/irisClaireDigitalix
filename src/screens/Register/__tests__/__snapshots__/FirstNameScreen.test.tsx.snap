// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirstNameScreen renders correctly and matches snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#2F2A5A",
      "flex": 1,
      "height": "100%",
      "justifyContent": "center",
    }
  }
  testID="first-name-screen"
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "column",
        "marginTop": 40,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "width": 300,
        }
      }
    >
      <Text
        style={
          {
            "color": "#dcdaee",
            "fontFamily": "mulishRegular",
            "fontSize": 18,
            "marginBottom": 20,
          }
        }
      >
        Quelle est votre prénom ?
      </Text>
    </View>
    <View
      style={
        {
          "marginBottom": 20,
          "width": 675,
        }
      }
    >
      <View
        style={
          {
            "paddingHorizontal": 10,
            "width": "100%",
          }
        }
        testID="RNE__Input__view-wrapper"
      >
        <View
          collapsable={false}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#EEEDF6",
              "borderBottomColor": "#CBA135",
              "borderBottomWidth": 1,
              "borderColor": "#86939e",
              "borderLeftColor": "#CBA135",
              "borderLeftWidth": 1,
              "borderRadius": 6,
              "flexDirection": "row",
              "padding": 3,
              "transform": [
                {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "marginVertical": 4,
                "paddingRight": 4,
              }
            }
          >
            <Icon
              name="user"
              size={28}
              style={
                {
                  "color": "#bc9531",
                  "marginLeft": 10,
                }
              }
            />
          </View>
          <TextInput
            editable={true}
            onChangeText={[Function]}
            placeholder="Saisissez votre prénom"
            placeholderTextColor="#B7B7E1"
            style={
              {
                "color": "#2F2A5A",
                "flex": 1,
                "fontFamily": "mulishRegular",
                "fontSize": 14,
                "marginLeft": 10,
                "minHeight": 40,
              }
            }
            testID="RNE__Input__text-input"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <Text
          style={
            {
              "color": "#ff190c",
              "fontSize": 12,
              "margin": 5,
            }
          }
        />
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "borderRadius": 16,
            "marginBottom": 10,
            "paddingVertical": 5,
          },
          {
            "backgroundColor": "#FF7648",
            "width": 637.5,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "mulishBold",
              "fontSize": 14,
            },
            {
              "color": "#eeedf6",
            },
          ]
        }
      >
        Suivant
      </Text>
    </View>
  </View>
</View>
`;
